diff --git a/tty-features.c b/tty-features.c
index 31d9b7a8..d7b01d3b 100644
--- a/tty-features.c
+++ b/tty-features.c
@@ -479,7 +479,7 @@ tty_default_features(int *feat, const char *name, u_int version)
 		},
 		{ .name = "tmux",
 		  .features = TTY_FEATURES_BASE_MODERN_XTERM
-			      ",ccolour,cstyle,focus,overline,usstyle,hyperlinks"
+			      ",ccolour,extkeys,cstyle,focus,overline,usstyle,hyperlinks"
 		},
 		{ .name = "rxvt-unicode",
 		  .features = "256,bpaste,ccolour,cstyle,mouse,title,ignorefkeys"
