diff --git a/src/lib/common/CMakeLists.txt b/src/lib/common/CMakeLists.txt
index cb4de1a..40e86fc 100644
--- a/src/lib/common/CMakeLists.txt
+++ b/src/lib/common/CMakeLists.txt
@@ -45,8 +45,8 @@ install(FILES ${PROJECT_BINARY_DIR}/man5/softhsm2.conf.5
         DESTINATION ${CMAKE_INSTALL_MANDIR}/man5
         )
 
-install(DIRECTORY DESTINATION ${CMAKE_INSTALL_LOCALSTATEDIR}/lib/softhsm/tokens)
+        install(DIRECTORY DESTINATION $ENV{DESTDIR}${CMAKE_INSTALL_LOCALSTATEDIR}/lib/softhsm/tokens)
 
 install(CODE
         "execute_process(
-        COMMAND chmod 1777 ${CMAKE_INSTALL_LOCALSTATEDIR}/lib/softhsm/tokens)")
+        COMMAND chmod 1777 \$ENV{DESTDIR}${CMAKE_INSTALL_LOCALSTATEDIR}/lib/softhsm/tokens)")
